File: src/type.c
#define CAML_NAME_SPACE
#include <caml/mlvalues.h>
#include <caml/fail.h>
#include <caml/memory.h>

#include "binaryen-c.h"
#include "ocaml_helpers.h"

CAMLprim value
caml_binaryen_type_none(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeNone();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_int32(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeInt32();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_int64(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeInt64();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_float32(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeFloat32();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_float64(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeFloat64();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_vec128(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeVec128();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_funcref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeFuncref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_anyref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeAnyref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_eqref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeEqref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_i31ref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeI31ref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_dataref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeDataref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_arrayref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeArrayref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_stringref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeStringref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_stringview_wtf8(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeStringviewWTF8();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_stringview_wtf16(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeStringviewWTF16();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_stringview_iter(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeStringviewIter();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_nullref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeNullref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_null_externref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeNullExternref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_null_funcref(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeNullFuncref();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_unreachable(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeUnreachable();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_auto(value unit) {
  CAMLparam1(unit);
  BinaryenType ty = BinaryenTypeAuto();
  CAMLreturn(alloc_BinaryenType(ty));
}

CAMLprim value
caml_binaryen_type_create(value tys) {
  CAMLparam1(tys);
  int len = array_length(tys);
  BinaryenType valueTypes[len];
  for (int i = 0; i < len; i++) {
    valueTypes[i] = BinaryenType_val(Field(tys, i));
  }
  BinaryenType newType = BinaryenTypeCreate(valueTypes, len);
  CAMLreturn(alloc_BinaryenType(newType));
}

CAMLprim value
caml_binaryen_type_expand(value _ty) {
  CAMLparam1(_ty);
  BinaryenType ty = BinaryenType_val(_ty);
  int arity = BinaryenTypeArity(ty);
  BinaryenType* buf = malloc(sizeof(BinaryenType) * arity);
  BinaryenTypeExpand(ty, buf);

  value typeArray = caml_alloc(arity, 0);
  for (int i = 0; i < arity; i++) {
    Field(typeArray, i) = alloc_BinaryenType(buf[i]);
  }

  free(buf);

  CAMLreturn(typeArray);
}

CAMLprim value
caml_binaryen_type_is_nullable(value _ty) {
  CAMLparam1(_ty);
  BinaryenType ty = BinaryenType_val(_ty);
  CAMLreturn(Val_bool(BinaryenTypeIsNullable(ty)));
}

CAMLprim value
caml_binaryen_type_from_heap_type(value _ty, value _nullable) {
  CAMLparam2(_ty, _nullable);
  BinaryenHeapType heapTy = BinaryenHeapType_val(_ty);
  bool nullable = Bool_val(_nullable);
  BinaryenType ty = BinaryenTypeFromHeapType(heapTy, nullable);
  CAMLreturn(alloc_BinaryenType(ty));
}

File: src/type.js
//Provides: caml_binaryen_type_none
//Requires: Binaryen
function caml_binaryen_type_none() {
  return Binaryen.Type.none;
}

//Provides: caml_binaryen_type_int32
//Requires: Binaryen
function caml_binaryen_type_int32() {
  return Binaryen.Type.i32;
}

//Provides: caml_binaryen_type_int64
//Requires: Binaryen
function caml_binaryen_type_int64() {
  return Binaryen.Type.i64;
}

//Provides: caml_binaryen_type_float32
//Requires: Binaryen
function caml_binaryen_type_float32() {
  return Binaryen.Type.f32;
}

//Provides: caml_binaryen_type_float64
//Requires: Binaryen
function caml_binaryen_type_float64() {
  return Binaryen.Type.f64;
}

//Provides: caml_binaryen_type_vec128
//Requires: Binaryen
function caml_binaryen_type_vec128() {
  return Binaryen.Type.v128;
}

//Provides: caml_binaryen_type_funcref
//Requires: Binaryen
function caml_binaryen_type_funcref() {
  return Binaryen.Type.funcref;
}

//Provides: caml_binaryen_type_anyref
//Requires: Binaryen
function caml_binaryen_type_anyref() {
  return Binaryen.Type.anyref;
}

//Provides: caml_binaryen_type_eqref
//Requires: Binaryen
function caml_binaryen_type_eqref() {
  return Binaryen.Type.eqref;
}

//Provides: caml_binaryen_type_i31ref
//Requires: Binaryen
function caml_binaryen_type_i31ref() {
  return Binaryen.Type.i31ref;
}

//Provides: caml_binaryen_type_structref
//Requires: Binaryen
function caml_binaryen_type_structref() {
  return Binaryen.Type.structref;
}

//Provides: caml_binaryen_type_arrayref
//Requires: Binaryen
function caml_binaryen_type_arrayref() {
  return Binaryen.Type.arrayref;
}

//Provides: caml_binaryen_type_stringref
//Requires: Binaryen
function caml_binaryen_type_stringref() {
  return Binaryen.Type.stringref;
}

//Provides: caml_binaryen_type_nullref
//Requires: Binaryen
function caml_binaryen_type_nullref() {
  return Binaryen.Type.nullref;
}

//Provides: caml_binaryen_type_null_externref
//Requires: Binaryen
function caml_binaryen_type_null_externref() {
  return Binaryen.Type.nullexternref;
}

//Provides: caml_binaryen_type_null_funcref
//Requires: Binaryen
function caml_binaryen_type_null_funcref() {
  return Binaryen.Type.nullfuncref;
}

//Provides: caml_binaryen_type_exnref
//Requires: Binaryen
function caml_binaryen_type_exnref() {
  return Binaryen.Type.exnref;
}

//Provides: caml_binaryen_type_null_exnref
//Requires: Binaryen
function caml_binaryen_type_null_exnref() {
  return Binaryen.Type.nullexnref;
}

//Provides: caml_binaryen_type_unreachable
//Requires: Binaryen
function caml_binaryen_type_unreachable() {
  return Binaryen.Type.unreachable;
}

//Provides: caml_binaryen_type_auto
//Requires: Binaryen
function caml_binaryen_type_auto() {
  return Binaryen.Type.auto;
}

//Provides: caml_binaryen_type_create
//Requires: Binaryen
//Requires: caml_js_from_array
function caml_binaryen_type_create(typs) {
  return Binaryen.createType(caml_js_from_array(typs));
}

//Provides: caml_binaryen_type_expand
//Requires: Binaryen
//Requires: caml_js_to_array
function caml_binaryen_type_expand(typ) {
  return caml_js_to_array(Binaryen.expandType(typ));
}

//Provides: caml_binaryen_type_is_nullable
//Requires: Binaryen
function caml_binaryen_type_is_nullable(typ) {
  return Binaryen._BinaryenTypeIsNullable(typ);
}

//Provides: caml_binaryen_type_from_heap_type
//Requires: Binaryen
function caml_binaryen_type_from_heap_type(typ, nullable) {
  return Binaryen._BinaryenTypeFromHeapType(typ, nullable);
}